File: src-tauri/src/store.rs
use std::path::PathBuf;
use std::sync::Arc;
use tauri::{AppHandle, Manager, Runtime};
use tauri_plugin_store::{Store, StoreExt};

pub fn get_global_store<R: Runtime>(app: &AppHandle<R>) -> Result<Arc<Store<R>>, String> {
    app.store("store.json").map_err(|e| e.to_string())
}

pub fn get_base_dir<R: Runtime>(app: &AppHandle<R>) -> Result<PathBuf, String> {
    let global = get_global_store(app).ok();
    if let Some(ref store) = global {
        if let Some(custom) = store
            .get("custom_base_path")
            .and_then(|v| v.as_str().map(|s| s.to_string()))
        {
            let p = PathBuf::from(custom);
            if p.is_absolute() {
                return Ok(p);
            }
        }
    }
    app.path().app_data_dir().map_err(|e| e.to_string())
}

pub fn normalize_domain(domain: &str) -> String {
    let mut domain = domain.to_lowercase();

    if !domain.contains("://") {
        let is_local = domain.starts_with("127.0.0.1")
            || domain.starts_with("localhost")
            || domain
                .split(':')
                .next()
                .unwrap_or("")
                .split('.')
                .all(|s| s.parse::<u32>().is_ok());

        let protocol = if is_local { "http://" } else { "https://" };
        domain = format!("{}{}", protocol, domain);
    }

    if let Ok(url) = url::Url::parse(&domain) {
        let mut origin = url.origin().ascii_serialization();

        if url.host_str() == Some("localhost") {
            origin = origin.replace("localhost", "127.0.0.1");
        }

        origin
    } else {
        domain.trim().to_lowercase()
    }
}

pub fn get_domain_folder<R: Runtime>(app: &AppHandle<R>, domain: &str) -> Result<PathBuf, String> {
    let normalized = normalize_domain(domain);
    let mut base = get_base_dir(app)?;
    let safe_domain = normalized
        .replace(|c: char| !c.is_alphanumeric(), "_")
        .to_lowercase();
    base.push("domains");
    base.push(safe_domain);
    Ok(base)
}

pub fn get_domain_store<R: Runtime>(
    app: &AppHandle<R>,
    domain: &str,
) -> Result<Arc<Store<R>>, String> {
    let normalized = normalize_domain(domain);
    let folder = get_domain_folder(app, &normalized)?;
    let _ = std::fs::create_dir_all(&folder);
    let store_path = folder.join("store.json");
    app.store(store_path).map_err(|e| e.to_string())
}

pub fn get_current_domain<R: Runtime>(app: &AppHandle<R>) -> Option<String> {
    let store = get_global_store(app).ok()?;
    store
        .get("domain")
        .and_then(|v| v.as_str().map(|s| s.to_string()))
}

pub fn get_current_store<R: Runtime>(app: &AppHandle<R>) -> Result<Arc<Store<R>>, String> {
    if let Some(domain) = get_current_domain(app) {
        let store = get_domain_store(app, &domain)?;
        if store.get("domain").is_none() {
            let _ = store.set("domain", serde_json::json!(domain));
            let _ = store.save();
        }
        Ok(store)
    } else {
        get_global_store(app)
    }
}

pub fn get_data_dir<R: Runtime>(app: &AppHandle<R>) -> Result<PathBuf, String> {
    if let Some(domain) = get_current_domain(app) {
        get_domain_folder(app, &domain)
    } else {
        get_base_dir(app)
    }
}

File: src/lib/store.ts
import { appDataDir, join } from "@tauri-apps/api/path";
import { mkdir } from "@tauri-apps/plugin-fs";
import { load, type Store } from "@tauri-apps/plugin-store";

let globalStore: Store | null = null;
const domainStores: Record<string, Store> = {};

export async function getGlobalStore(): Promise<Store> {
    if (!globalStore) {
        globalStore = await load("store.json");
    }
    return globalStore;
}

export function normalizeDomain(domain: string): string {
    try {
        const hasProtocol = domain.includes("://");
        let protocol = "https://";

        const plainDomain = hasProtocol ? domain.split("://")[1] : domain;
        if (
            plainDomain.startsWith("127.0.0.1") ||
            plainDomain.startsWith("localhost") ||
            plainDomain
                .split(":")[0]
                .split(".")
                .every((part) => !isNaN(Number(part)))
        ) {
            protocol = "http://";
        }

        const url = new URL(hasProtocol ? domain : `${protocol}${domain}`);
        let origin = url.origin.toLowerCase();

        if (url.hostname === "localhost") {
            origin = origin.replace("localhost", "127.0.0.1");
        }

        return origin;
    } catch (e) {
        return domain.toLowerCase().trim();
    }
}

export async function getDomainFolder(domain: string): Promise<string> {
    const normalized = normalizeDomain(domain);
    const base = await appDataDir();
    const folderName = normalized.replace(/[^a-z0-9]/gi, "_").toLowerCase();
    const domainFolder = await join(base, "domains", folderName);
    return domainFolder;
}

export async function getDomainStore(domain: string): Promise<Store> {
    const normalized = normalizeDomain(domain);
    if (!domainStores[normalized]) {
        const folder = await getDomainFolder(normalized);
        try {
            await mkdir(folder, { recursive: true });
        } catch (e) { }
        domainStores[normalized] = await load(await join(folder, "store.json"));
    }
    return domainStores[normalized];
}

export async function getCurrentDomain(): Promise<string | null> {
    const store = await getGlobalStore();
    const domain = await store.get<string>("domain");
    return domain ?? null;
}

export async function getStore(): Promise<Store> {
    const domain = await getCurrentDomain();
    if (domain) {
        return await getDomainStore(domain);
    }
    return await getGlobalStore();
}

export async function getSavedDomains(): Promise<string[]> {
    const store = await getGlobalStore();
    return (await store.get<string[]>("domains")) || [];
}

export async function addSavedDomain(domain: string) {
    const normalized = normalizeDomain(domain);
    const store = await getGlobalStore();
    const domains = await getSavedDomains();
    if (!domains.includes(normalized)) {
        domains.push(normalized);
        await store.set("domains", domains);
        await store.save();
    }
}
